drv/hisi_sec.c

```diff
@@ -756,3 +759,304 @@ static struct wd_digest_driver hisi_digest_driver = {
+
+static int fill_aead_bd2_alg(struct wd_aead_msg *msg,
+		struct hisi_sec_sqe *sqe)
+{
+	__u8 c_key_len = 0;
+	__u32 d_alg = 0;
+	int ret = 0;
+
+	switch (msg->calg) {
+	case WD_CIPHER_SM4:
+		sqe->type2.c_alg = C_ALG_SM4;
+		sqe->type2.icvw_kmode = CKEY_LEN_SM4 << SEC_CKEY_OFFSET;
+		break;
+	case WD_CIPHER_AES:
+		sqe->type2.c_alg = C_ALG_AES;
+		if (msg->ckey_bytes == AES_KEYSIZE_128) {
+			c_key_len = CKEY_LEN_128BIT;
+		} else if (msg->ckey_bytes == AES_KEYSIZE_192) {
+			c_key_len = CKEY_LEN_192BIT;
+		} else if (msg->ckey_bytes == AES_KEYSIZE_256) {
+			c_key_len = CKEY_LEN_256BIT;
+		} else {
+			WD_ERR("failed to check AES key size!\n");
+			return -EINVAL;
+		}
+		sqe->type2.icvw_kmode = (__u16)c_key_len << SEC_CKEY_OFFSET;
+		break;
+	default:
+		WD_ERR("failed to check aead calg type!\n");
+		ret = -EINVAL;
+	}
+
+	/* CCM/GCM this region is set to 0 */
+	if (msg->cmode == WD_CIPHER_CCM ||
+	    msg->cmode == WD_CIPHER_GCM)
+		return ret;
+
+	if (unlikely(msg->auth_bytes & WORD_ALIGNMENT_MASK)) {
+		WD_ERR("failed to check aead auth_bytes!\n");
+		return -WD_EINVAL;
+	}
+	sqe->type2.mac_key_alg = msg->auth_bytes / WORD_BYTES;
+
+	if (unlikely(msg->akey_bytes & WORD_ALIGNMENT_MASK)) {
+		WD_ERR("failed to check aead auth key bytes!\n");
+		return -WD_EINVAL;
+	}
+	sqe->type2.mac_key_alg |= (__u32)(msg->akey_bytes /
+		WORD_BYTES) << MAC_LEN_OFFSET;
+
+	switch (msg->dalg) {
+	case WD_DIGEST_SHA1:
+		d_alg = A_ALG_HMAC_SHA1 << AUTH_ALG_OFFSET;
+		break;
+	case WD_DIGEST_SHA256:
+		d_alg = A_ALG_HMAC_SHA256 << AUTH_ALG_OFFSET;
+		break;
+	case WD_DIGEST_SHA512:
+		d_alg = A_ALG_HMAC_SHA512 << AUTH_ALG_OFFSET;
+		break;
+	default:
+		WD_ERR("failed to check aead dalg type!\n");
+		ret = -WD_EINVAL;
+	}
+	sqe->type2.mac_key_alg |= d_alg;
+
+	return ret;
+}
+
+static int fill_aead_bd2_mode(struct wd_aead_msg *msg,
+		struct hisi_sec_sqe *sqe)
+{
+	__u16 c_mode;
+
+	switch (msg->cmode) {
+	case WD_CIPHER_ECB:
+		c_mode = C_MODE_ECB;
+		break;
+	case WD_CIPHER_CBC:
+		c_mode = C_MODE_CBC;
+		break;
+	case WD_CIPHER_CTR:
+		c_mode = C_MODE_CTR;
+		break;
+	case WD_CIPHER_CCM:
+		c_mode = C_MODE_CCM;
+		sqe->type_auth_cipher |= NO_AUTH << SEC_AUTH_OFFSET;
+		sqe->type2.alen_ivllen = msg->assoc_bytes;
+		sqe->type2.icvw_kmode |= msg->auth_bytes;
+		break;
+	case WD_CIPHER_GCM:
+		c_mode = C_MODE_GCM;
+		sqe->type_auth_cipher |= NO_AUTH << SEC_AUTH_OFFSET;
+		sqe->type2.alen_ivllen = msg->assoc_bytes;
+		sqe->type2.icvw_kmode |= msg->auth_bytes;
+		break;
+	default:
+		WD_ERR("failed to check aead cmode type!\n");
+		return -EINVAL;
+	}
+	sqe->type2.icvw_kmode |= (__u16)(c_mode) << SEC_CMODE_OFFSET;
+
+	return 0;
+}
+
+static void set_aead_auth_iv(struct wd_aead_msg *msg)
+{
+#define IV_LAST_BYTE1		1
+#define IV_LAST_BYTE2		2
+#define IV_CTR_INIT		1
+#define IV_CM_CAL_NUM		2
+#define IV_CL_MASK		0x7
+#define IV_FLAGS_OFFSET	0x6
+#define IV_CM_OFFSET		0x3
+#define IV_LAST_BYTE2_MASK	0xFF00
+#define IV_LAST_BYTE1_MASK	0xFF
```

Comment:
please move these at the head of file

Reply:
There is no need to move to the file header，them just used in this function